uspace/lib/cpp/include/__bits/thread/shared_state.hpp

```
@@ -0,0 +1,409 @@
+                 */
+                if (!this->is_set())
+                    invoke_(make_index_sequence<sizeof...(Args)>{});
+            }
+
+            ~deferred_shared_state() override
+            {
+                destroy();
+            }
+
+        protected:
+            function<R(decay_t<Args>...)> func_;
+            tuple<decay_t<Args>...> args_;
+
+            template<size_t... Is>
+            void invoke_(index_sequence<Is...>)
+            {
+                try
+                {
+                    if constexpr (!is_same_v<R, void>)
+                        this->set_value(invoke(move(func_), get<Is>(move(args_))...));
+                    else
+                    {
+                        invoke(move(func_), get<Is>(move(args_))...);
+                        this->mark_set(true);
+                    }
+                }
+                catch(const exception& __exception)
+                {
+                    this->set_exception(make_exception_ptr(__exception));
+                }
+            }
+
+            future_status timed_wait_(aux::time_unit_t) override
+            {
+                /**
+                 * Note: Neither of the wait_ functions has any effect
+                 *       for deferred functions spawned by async (which
+                 *       are the only users of this state type).
+                 */
+                return future_status::deferred;
+            }
+    };
+
+    /**
+     * Note: The following two functions should:
+     *   1) Increment refcount.
+     *   2) Store ptr to a vector of shared_state_base ptrs
+     *      (as those have ::mark_set member functions).
+     *   3) If not done already, register a function
+     *      executing all these in the thread_atexit function
+     *      once that is implemented.
+     */
+
+    template<class R>
+    void set_state_value_at_thread_exit(shared_state<R>* state)
+    {
+        // TODO: implement
```

Probably should abort/throw an exception instead of silently doing nothing.